solr/src/main/java/com/yahoo/ycsb/db/SolrClient.java

```diff
@@ -0,0 +1,316 @@
+    // TO-DO: Add a check that the collection exists
+    // Check if Solr cluster is running in SolrCloud mode
+    cloudMode = Boolean.parseBoolean(props.getProperty("solr.cloud", DEFAULT_CLOUD_MODE));
+    System.out.println("Solr Cloud Mode = " + cloudMode);
+    if (cloudMode) {
+      System.out.println("Solr Zookeeper Remote Hosts = "
+          + props.getProperty("zookeeper.host.list", DEFAULT_REMOTE_ZOOKEEPER_HOST));
+
+      CloudSolrClient server = new CloudSolrClient(
+          props.getProperty("zookeeper.host.list", DEFAULT_REMOTE_ZOOKEEPER_HOST));
+      server.setDefaultCollection(collectionName);
+      cloudClient = server;
+    } else { // Start node only if cluster mode is disabled
+      HttpSolrClient server = new HttpSolrClient(
+          "http://" + props.getProperty("solr.host.list", DEFAULT_REMOTE_SOLR_HOST) + "/solr");
+      httpClient = server;
+    }
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    try {
+      if (cloudMode) {
+        cloudClient.close();
+      } else {
+        httpClient.close();
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Insert a record in the database. Any field/value pairs in the specified values HashMap will be
+   * written into the record with the specified record key.
+   *
+   * @param table
+   *          The name of the table
+   * @param key
+   *          The record key of the record to insert.
+   * @param values
+   *          A HashMap of field/value pairs to insert in the record
+   * @return Zero on success, a non-zero error code on error. See this class's description for a
+   *         discussion of error codes.
+   */
+  @Override
+  public Status insert(String table, String key, HashMap<String, ByteIterator> values) {
+    try {
+      SolrInputDocument doc = new SolrInputDocument();
+
+      doc.addField("id", key);
+      for (Entry<String, String> entry : StringByteIterator.getStringMap(values).entrySet()) {
+        doc.addField(entry.getKey(), entry.getValue());
+      }
+
+      if (cloudMode) {
+        cloudClient.add(table, doc);
+        cloudClient.commit(table);
+      } else {
+        httpClient.add(table, doc);
+        httpClient.commit(table);
+      }
+
+      return Status.OK;
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+    return Status.ERROR;
+  }
+
+  /**
+   * Delete a record from the database.
+   *
+   * @param table
+   *          The name of the table
+   * @param key
+   *          The record key of the record to delete.
+   * @return Zero on success, a non-zero error code on error. See this class's description for a
+   *         discussion of error codes.
+   */
+  @Override
+  public Status delete(String table, String key) {
+    try {
+      UpdateResponse res = (cloudMode) ? cloudClient.deleteById(table, key, 1000)
+          : httpClient.deleteById(table, key, 1000);
+      if (res.getStatus() > 0) {
+        throw new Exception("");
```

Comment:
There might be a better way to dealing with bad states here. If res.getStatus() gives back any useful information, you could return different status codes that have more meaning than Status.ERROR. There are some predefined status codes here: https://github.com/brianfrankcooper/YCSB/blob/master/core/src/main/java/com/yahoo/ycsb/Status.java, you can also create more specific ones if necessary.